Build failure: gdal ([javac] error: Source option 7 is no longer supported. Use 8 or later)

### Steps To Reproduce

Steps to reproduce the behavior:
1. build *gdal*

### Build log

```
gdal> java/overview-tree.html
gdal> [100%] Generating gdal.jar, build/maven/gdal-3.8.4.jar, build/maven/gdal-3.8.4-sources.jar, build/maven/gdal-3.8.4.pom, build/maven/gdal-3.8.4-javadoc.jar
gdal> Buildfile: /build/source/swig/java/build.xml
gdal>
gdal> init:
gdal>      [echo] version = 3.8.4
gdal>
gdal> compile:
gdal>     [mkdir] Created dir: /build/source/build/swig/java/build/classes
gdal>     [javac] Compiling 88 source files to /build/source/build/swig/java/build/classes
gdal>     [javac] warning: [options] bootstrap class path not set in conjunction with -source 7
gdal>     [javac] error: Source option 7 is no longer supported. Use 8 or later.
gdal>     [javac] error: Target option 7 is no longer supported. Use 8 or later.
gdal>
gdal> BUILD FAILED
gdal> /build/source/swig/java/build.xml:25: Compile failed; see the compiler error output for details.
gdal>
gdal> Total time: 0 seconds
gdal> make[2]: *** [swig/java/CMakeFiles/java_binding.dir/build.make:107: swig/java/gdal.jar] Error 1
gdal> make[1]: *** [CMakeFiles/Makefile2:10036: swig/java/CMakeFiles/java_binding.dir/all] Error 2
gdal> make[1]: *** Waiting for unfinished jobs....
```

### Notify maintainers

<!--
Please @ people who are in the `meta.maintainers` list of the offending package or module.
If in doubt, check `git blame` for whoever last touched something.
-->

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
output here
```
this path will be fetched (0.00 MiB download, 0.00 MiB unpacked):
  /nix/store/imd120fan53pgy6qi6kra0q15v5kn6dn-nix-info
copying path '/nix/store/imd120fan53pgy6qi6kra0q15v5kn6dn-nix-info' from 'https://cache.nixos.org'...
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.80, NixOS, 23.11 (Tapir), 23.11.20240305.880992d`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.1`
 - nixpkgs: `/home/imincik/.cache/nixpkgs-review/pr-294874/nixpkgs`
---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/development/libraries/gdal/default.nix
+++ b/pkgs/development/libraries/gdal/default.nix
@@ -2,6 +2,7 @@
 , stdenv
 , callPackage
 , fetchFromGitHub
+, fetchpatch
 
 , useMinimalFeatures ? false
 , useTiledb ? (!useMinimalFeatures) && !(stdenv.isDarwin && stdenv.isx86_64)
@@ -87,6 +88,14 @@ stdenv.mkDerivation (finalAttrs: {
     rev = "v${finalAttrs.version}";
     hash = "sha256-R9VLof13OXPbWGHOG1Q4WZWSPoF739C6WuNWxoIwKTw=";
   };
+
+  patches = [
+    # bump java source option to fix build with JDK 21
+    (fetchpatch {
+      url = "https://github.com/OSGeo/gdal/commit/ca2eb4130750b0e6365f738a5f8ff77081f5c5bb.patch";
+      sha256 = "sha256-wShYm9yA7twJR72co+Tvf/IuYXqbI0OrjWl0uqC3bwo=";
+    })
+  ];
 
   nativeBuildInputs = [
     bison

